src/styles/createGenerateClassName.js

```diff
@@ -36,12 +37,17 @@ export default function createGenerateClassName() {
       ].join(''),
     );
 
-    if (process.env.NODE_ENV === 'production') {
+    if (process.env.NODE_ENV === 'production' && !dangerouslyUseGlobalCSS) {
       return `c${ruleCounter}`;
     }
 
     if (styleSheet && styleSheet.options.meta) {
       let meta = styleSheet.options.meta;
+
+      if (dangerouslyUseGlobalCSS && meta.match(/^Mui/)) {
+        return `${meta}-${rule.key}`;
+      }
+
       // Sanitize the string as will be used in development to prefix the generated
       // class name.
       meta = meta.replace(new RegExp(/[!"#$%&'()*+,./:; <=>?@[\\\]^`{|}~]/g), '-');
```

Comment:
Shouldn't this be done also when dangerouslyUseGlobalCSS is true?

Reply:
No need. We use the name of the stylesheet e.g. `MuiButton` and the name of the key e.g `root`. They will never be anything fancy. If we have, It's better to surface the issue.